bug(Stock Value report crashes server with too much data)

lomamech opened the pull request:
- Improvement and optimization of the stock value by deposit report

closes #4116

lomamech committed:
Saving changes

lomamech committed:
bug(Stock Value report crashes server with too much data)
- Improvement and optimization of the stock value by deposit report

closes #4116

jniles commented on migration-hev_mai_2020.sql:
```diff
@@ -0,0 +1,9 @@
+SET character_set_database = 'utf8mb4';
+SET collation_database = 'utf8mb4_unicode_ci';
+SET CHARACTER SET utf8mb4, CHARACTER_SET_CONNECTION = utf8mb4;
+SET collation_connection = 'utf8mb4_unicode_ci';
```
Is this really necessary?  The SQL dump should be exporting with these defaults. Is that not the case?

lomamech replied:
> Is this really necessary? The SQL dump should be exporting with these defaults. Is that not the case?

This portion of code was added when I rebase with master, I haven't added it personally

jniles commented on server/controllers/stock/reports/stock/value.js:
```diff
@@ -30,21 +30,109 @@ async function reporting(_options, session) {
+   * but also the cumulative value of entries *quantityCum* and exits *cumOut*
+   * and the quantity in stock after each movement of stock *quantityStock*
+  */
+  const stockValues = await db.exec(sqlGetInventories, [db.bid(options.depot_uuid), options.dateTo]);
```
Sp, if I am understanding this correctly, this code will generate `1 + N` SQL queries in a transaction, where `N` is the number of inventory items in a depot.  At Vanga, the total number of inventory items is currently 784.  At IMCK, the total number is 1302.  So we could potentially get hundreds of SQL queries launched from a single query here.

... I'm not sure how optimal this is in low-end hardware like a Raspberry Pi.

lomamech replied:
the way I understood the logic of the stock value report came down to retrieving the last line of the stock card report, my idea was to launch several requests on the stock movement table in order to calculate for each unit cost weight and the stock value after each transaction

lomamech replied:
I had given up on the idea of using an aggregation table because of the risk of having backdated stock movements

jniles commented on server/controllers/stock/reports/stock/value.js:
```diff
@@ -30,21 +30,109 @@ async function reporting(_options, session) {
+    });
+  });
+
+  let stockTolal = 0;
```
This looks like a typo - it should be "stockTotal"

jniles commented on server/controllers/stock/reports/stock_value.report.handlebars:
```diff
@@ -4,61 +4,56 @@
+              <tr>
+                <td style="width:50%">{{inventory_name}}</td>
+                <td class="text-right">{{stockQtt}}</td>
+                <td class="text-right">{{currency (multiply stockUnitCost ../rate) ../currency_id}}</td>
```
Can you do this multiplication before putting it in the template?

jniles commented on server/controllers/stock/reports/stock_value.report.handlebars:
```diff
@@ -4,61 +4,56 @@
+              </tr>
+            {{/each}}
+            <!-- no data -->
+            {{#if emptyResult}}
```
There is no need for this "emptyResult" check.  Just use `{{#each}}{{else}}{{/each}}` as discussed in WhatsApp.

So, it would look like this:
```hbs
{{#each stockValues}}
   {{! ... etc ... }}
{{else}}
  <tr><th class="text-center" colspan="4">{{translate 'STOCK.NO_DATA'}}</th></tr>
{{/each}}
```

jniles commented on server/controllers/stock/reports/stock/value.js:
```diff
@@ -30,21 +30,109 @@ async function reporting(_options, session) {
   const options = (typeof (_options.params) === 'string') ? JSON.parse(_options.params) : _options.params;
   data.dateTo = options.dateTo;
   data.depot = await db.one('SELECT * FROM depot WHERE uuid=?', [db.bid(options.depot_uuid)]);
-  const stockValues = await db.exec('CALL stockValue(?,?,?);', [
```
If we are not longer using the `stockValue()` procedure, let's remove it from the stored procedures.

jniles commented on server/controllers/stock/reports/stock/value.js:
```diff
@@ -30,21 +30,109 @@ async function reporting(_options, session) {
 
   const stockValueElements = options.exclude_zero_value
-    ? stockValues[0].filter(item => item.stockValue > 0) : stockValues[0];
+    ? stockValues.filter(item => item.stockValue > 0) : stockValues;
```
Is there anyway we can apply this filter higher in logic?  Preferably in the SQL.  It would probably improve performance.

lomamech replied:
it will be difficult seeing that the products we see the stock shortage of a product after on the last line, after having noted all the entries and exits from Stock

jniles reviewed (commented):
I tried to test this, but it still crashes for me.  My virtual machine only has 4GB of RAM, but that should be more than enough.  After 2 minutes, the request times out and retries.  It also pegs my MySQL CPU usage at over 300%.

![image](https://user-images.githubusercontent.com/896472/83120668-3e298300-a0c9-11ea-855c-863717da05f1.png)

I'm not sure this solves the problem at all.  I still think, as mentioned in our previous discussion, that something like `period_total` is the way to go.  That way, we are doing very little computation in the actual report itself.

![caKfGd60mv](https://user-images.githubusercontent.com/896472/83120852-7af57a00-a0c9-11ea-9d20-70151a94761b.gif)

lomamech committed:
Remove not necessary migration-hev_mai_2020.sql in Master

lomamech committed:
Save Change

mbayopanda reviewed (commented):
![image](https://user-images.githubusercontent.com/5445251/83377678-4423aa80-a3ce-11ea-8262-854c367fc67e.png)

it seems like stock_sheet report is not available

lomamech committed:
improve(Stock Value Report)
- Improvement and optimization of the stock value report, reduction of
  operations on the MYSQL side

lomamech committed:
Sanitaze code

jniles reviewed (approved):
This is much better for performance.  The query takes 10 seconds on my machine (22 seconds total time).  It also doesn't peg MySQL at 300% CPU usage.  See below.
 
![8wged7bEs0](https://user-images.githubusercontent.com/896472/83609817-ac14f500-a576-11ea-80a6-1c1673a388c4.gif)

I still think we can do better via `period_total` style table, but this fixes the bug in production to an acceptable level where the report is usable.  Great work!

bors r+
